apps/internal/base/internal/storage/storage.go

```diff
@@ -218,18 +213,13 @@ func (m *Manager) Write(authParameters authority.AuthParams, tokenResponse acces
 	return account, nil
 }
 
-// Contract returns the Contract for read operations.
-func (m *Manager) Contract() *Contract {
-	return m.contract.Load().(*Contract)
-}
-
 func (m *Manager) getMetadataEntry(ctx context.Context, authorityInfo authority.Info) (authority.InstanceDiscoveryMetadata, error) {
-	m.cacheMu.Lock()
-	defer m.cacheMu.Unlock()
-
+	m.aadCacheMu.RLock()
```

Comment:
Cannot use `defer m.aadCacheMu.RUnlock()` here ?

Reply:
No, m.aadMetadata() takes the write lock and RWMutext doesn't support lock promotion.  I've added a comment explaining this.

Reply:
Thanks for the comment ! Makes sense.

Reply:
Sorry to unresolve this:

I'm looking at this and it looks like to me we could do the following to make this honor defers, put the locks in a single place and generally clean it up.  I think Joel this is how it would probably look had you been writing it fresh instead of this re-write.

```go
func (m *Manager) addMetadaFromCache(authorityInfo authority.Info) (authority.InstanceDiscoveryMetadata, error) {
  m.addCacheMu.RLock()
  defer m.addCacheMu.RUnlock()

  metadata, ok := m.aadCache[authorityInfo.Host]
  if ok {
    return metadata, nil
  }
  return metadata, fmt.Errorf("not found")
}

func (m *Manager) aadMetadata(ctx context.Context, authorityInfo authority.Info) (authority.InstanceDiscoveryMetadata, error) {
  if md, err := addMetadaFromCache(authorityInfo); err != nil {
    return md, nil
  }
  ...
}
```

Now in getMetadataEntry(), this should end up being:

```go
  metadata, err := m.aadMetadata(ctx, authorityInfo)
  ...
```
No more locks at that level, everything is deferred

Reply:
Fixed